### DISCUSSION TITLE
Send By WhatasAPP OR FOR WhatasAPP using LHC

### DISCUSSION THREAD

**Discussion by AppDesv** [LiveHelperChat/livehelperchat#1984] [General]
Hi Remdex,

I would like to ask a question and request your valuable assistance once again, as you always provide it in a kind and objective manner, even amid numerous questions that have already been clarified.

The issue I wish to address is as follows: Is it possible to create a new chat to initiate a conversation via WhatsApp using the API /restapi/chat? Currently, I am facing a limitation that only allows me to start conversations with WhatsApp numbers that have been previously attended to (reopening them and sending messages through /restapi/setchatstatus and /restapi/addmsguser). However, my goal is to initiate conversations with the numbers in my client database, aiming to provide more effective customer service.

When using the /restapi/chat call to create a new chat, I observe that the WhatsApp plugin (represented by the puzzle piece icon) is not activated, even after providing all necessary information such as the phone number and client ID.

Could you please guide me through the steps I should take to resolve this issue?

Thank you in advance for your invaluable assistance and guidance.

**Comment by remdex** [comment]
Few things. I assume you are using https://github.com/LiveHelperChat/fbmessenger extension.

1. You can send direct messages only in first 24 hours once you made contact.
2. You can send to send template at any time.

A Option, required development
What you need is `extension/fbmessenger/modules/lhfbwhatsapp/send.php` similar function what this does. `/site_admin/fbwhatsapp/send`
So you can just create a small extension, add Rest API call validation. Once initial message (whatsapp template) is send I store it as a send message not related to LHC. LHC Chat itself is created only once I receive incoming webhook call from customer in first his response... 

B Option, does not require development.
"/site_admin/webhooks/incomingwebhooks" you could also try to fake It. Create rest API which would make request to that call... And afterward just add normal message to chat. That flow also would work as lhc would see as external whatsapp chat. Sample payload you would find in "lh_chat_incoming" this table. After that you could send to that chat a command (template) knowing that customer never made a contact. 

`!fbtemplate {"template_name":"hello_world","template_lang":"en_us","args":{}}`

https://github.com/LiveHelperChat/fbmessenger#whatsapp-configuration

In any case it's just hour or two of work.

**Reply by AppDesv** [reply]
Thank you for the information regarding the use of the https://github.com/LiveHelperChat/fbmessenger extension. However, I would like to clarify that I am currently using Open-WA instead of the mentioned extension.

Do you think it would be a good idea to migrate to fbmessenger? I'm curious to know if this would have a significant impact on costs. While I am currently using Open-WA due to its open-source nature and accessibility, I am open to exploring other options if you believe it's worth it.

Regarding the costs associated with the fbmessenger extension, I am interested in obtaining more information on this matter. If there are any fees involved, could you provide details about the pricing?

**Comment by AppDesv** [comment]
Thank you for the information regarding the use of the https://github.com/LiveHelperChat/fbmessenger extension. However, I would like to clarify that I am currently using Open-WA instead of the mentioned extension.

Do you think it would be a good idea to migrate to fbmessenger? I'm curious to know if this would have a significant impact on costs. While I am currently using Open-WA due to its open-source nature and accessibility, I am open to exploring other options if you believe it's worth it.

Regarding the costs associated with the fbmessenger extension, I am interested in obtaining more information on this matter. If there are any fees involved, could you provide details about the pricing?

**Reply by remdex** [reply]
As you are using open-wa integration. B option is still valid. You fake incoming webhook, afterwards find that chat and using same addmsguser api call add message to it. All should work that way. 

To simplify "find that chat" incoming webhook configuration could be extended to return chat id instantly. Just that addidional job. For begining you could just use getchatlist after your call. Have in mind it terminates fast cgi process by default and does not wait untill requestis finished so you might meed to add some delay after that call...

**Comment by AppDesv** [comment]
You're welcome, Remdex! I'll try to implement it here :D. And I'll give you feedback!

I would greatly appreciate your opinion regarding a potential migration from Open-WA to the official FB. Do you think this change would be a wise decision?

My experience with Open-WA has been extremely positive, and it has served me satisfactorily. However, your initial observation, where you mentioned the possibility of me using the official FB, has prompted me to reflect on the advantages that this migration could bring.

**Reply by remdex** [reply]
In short fbmessenger has campaigns and fb whatsapp templates support, but requires official app, it's review and verification. Besides that there is not much of advantages.

If open-wa works fine for you I would stick with it just :) and B flow should work just fine :) In incoming webhook definition you can also define what should happen when new message arrives and old chat is closed - should be created always or old one opened on customer message :)

**Reply by GuilhermeDeLucca** [reply]
In addition to the higher cost, I am correct?

**Reply by AppDesv** [reply]
Does FBMessenger actually have a higher cost compared to openWA, Remdex?

I can't say for sure, could you confirm this for me, please?

**Reply by remdex** [reply]
I'm not fb expert. If maintenance is a cost for sure fb requires more attention. You should make your own analysis. Or ask chatgtp :D

**Reply by AppDesv** [reply]
Thank you for always being guided on how to proceed, believe it or not, your opinion is more valid than chatgtp. :D

**Comment by AppDesv** [comment]
Following this idea:

Option B, does not require development.
You could also try faking it by calling "/site_admin/webhooks/incomingwebhooks". Create a REST API that makes a request to that endpoint, and then simply add a regular message to the chat. This flow would also function, as Live Helper Chat (LHC) would perceive it as an external WhatsApp chat. You can find a sample payload in the "lh_chat_incoming" table. Afterward, you could send a command (template) to that chat, knowing that the customer never initiated contact.

Do I have to call "/site_admin/webhooks/incomingwebhooks" to create a new chat? Sorry if I didn't quite grasp the idea correctly!

**Comment by AppDesv** [comment]
Would it be something like this?

URL REST:

https://mylhc.com/site_admin/webhooks/incomingwebhooks

{
  "ignore_required": true,
  "ignore_bot": false,
  "department": [
    1
  ],
  "fields": {
    "Question": "Visitor Initial Message. If you are passing messages argument makes sense just skip this",
    "Username": "Visitor nick"
  },
  "chat_variables": {
    "UUID": "Just some internal variable"
  },
  "main_attr": {
    "remarks": "This is main attribute update"
  },
  "additional_data": [
    {
      "h": true,
      "identifier": "age",
      "key": "Age",
      "value": 18
    }
  ],
  "messages": [
    {
      "msg": "Visitor initial message",
      "time": 1691974335,
      "user_id": 0
    },
    {
      "msg": "Bot message",
      "time": 1691974335,
      "user_id": -2
    },
    {
      "msg": "System message",
      "time": 1691974335,
      "user_id": -1
    }
  ]
}

I know it's not this, but I still didn't understand the concept!

**Reply by remdex** [reply]
No, incomingwebhooks/webhooks/identifier it's same url you see while editing webhook itself. Endpoint url is presented in edit window. You should be familiar with it as you had to point it in open-wa confoguration.

Payload sample you will fond in incoming chat table ss I wrote before... it has to
Be open-wa payload structure

**Reply by AppDesv** [reply]
I'm doing it successfully like this

URL: https://mylhc.com/site_admin/restapi/chat

```
{
"ignore_required": true,
"ignore_bot": false,
"department": [1],
"fields": {
"Question": "Visitor Initial Message. If you are passing messages argument makes sense just skip this",
"Username": "Visitor nick"
},
"chat_variables": {
"UUID": "Just some internal variable"
},
"main_attr": {
"remarks": "This is main attribute update"
},
"additional_data": [
{
"h": true,
"identifier": "age",
"key": "Age",
"value": 18
}
],
"messages": [
{
"msg": "Visitor initial message",
"time": 1691973690,
"user_id": 0
},
{
"msg": "Bot message",
"time": 1691973690,
"user_id": -2
},
{
"msg": "System message",
"time": 1691973690,
"user_id": -1
}
]
}

```
Correctly creates the chat (but not linked to WhatsApp)

I tried to do it like this

URL: https://mylhc.com/site_admin/webhooks/incoming/open-wa

```
{
"ignore_required": true,
"ignore_bot": false,
"department": [1],
"fields": {
"Question": "Visitor Initial Message. If you are passing messages argument makes sense just skip this",
"Username": "Visitor nick"
},
"chat_variables": {
"UUID": "Just some internal variable"
},
"main_attr": {
"remarks": "This is main attribute update"
},
"additional_data": [
{
"h": true,
"identifier": "age",
"key": "Age",
"value": 18
}
],
"messages": [
{
"msg": "Visitor initial message",
"time": 1691973690,
"user_id": 0
},
{
"msg": "Bot message",
"time": 1691973690,
"user_id": -2
},
{
"msg": "System message",
"time": 1691973690,
"user_id": -1
}
]
}
```

Did not create the service

My WhatsApp works correctly on this URL

https://mylhc.com/site_admin/webhooks/incoming/open-wa

**Comment by AppDesv** [comment]
@remdex, sorry to bother you again. But I still don't understand how I make the call! I've been trying all day and nothing :/ Could you give a little more tangible explosion?

**Reply by remdex** [reply]
![image](https://github.com/LiveHelperChat/livehelperchat/assets/1146085/f2c3a509-da18-4eb4-ab6c-7333c51e1092)

From table `lh_chat_incoming` find payload and send same payload to your webhok URL from first screenshot.

Afterwards just modify payload to mach whom you are sending it. After that find that new chat through Rest API calls and just add message via Rest API calls also.

```
curl -X POST https://dummy.livehelperchat.com/webhooks/incoming/open-wa
     -H 'Content-Type: application/json' 
     -d '{"destination":......'
```

**Comment by AppDesv** [comment]
this is exactly the part I'm wondering about

-d '{"destination":......'


I already tried to send it in several ways

{
"ignore_required": true,
"ignore_bot": false,
"department": [
1
],
"fields": {
"Question": "Visitor Initial Message. If you are passing messages argument makes sense just skip this",
"Username": "Visitor nick"
},
"chat_variables": {
"UUID": "Just some internal variable"
},
"main_attr": {
"remarks": "This is main attribute update"
},
"additional_data": [
{
"h": true,
"identifier": "age",
"key": "Age",
"value": 18
}
],
"messages": [
{
"msg": "Visitor initial message",
"time": 1691974335,
"user_id": 0
},
{
"msg": "Bot message",
"time": 1691974335,
"user_id": -2
},
{
"msg": "System message",
"time": 1691974335,
"user_id": -1
}
]
}

OR

{
  "args": {
    "to": "00000000000@c.us or 00000000000-111111111@g.us",
    "content": "Hello World!"
  }
}

I feel really dumb @remdex, sorry for not understanding

**Reply by remdex** [reply]
Read few more times my replies. Sample payloads you can find in table i mentioned before

**Reply by AppDesv** [reply]
I'm totally lost now, very confused LOL ahahah

**Comment by AppDesv** [comment]
Rereading the msg here, looking at the base, are you suggesting that I recreate a duplicate call to the table row would that be it?

If so, why don't I simply reopen?

I want to create a new message (it would be a proactive invitation)

a msg that the visitor does not start would be initiated by the LHC

I'm sorry I'm so dumb @remdex

**Reply by remdex** [reply]
At the end it would
Be duplicate, but you can change body. In general up to you how you do that.

**Comment by AppDesv** [comment]
Am I going to insert this value through a sql query?

because I looked at the table that told me lh_chat_incoming

I still don't understand your idea, how to start a chat to appear on LCH

I can send direct msg to whatsAPP however I wanted it to be linked to the LCH I still don't understand or I don't know if @remdex you didn't understand me

**Reply by remdex** [reply]
Sorry, but sql inserting is a wrong case. I can't explain anyway more detailed than I already did... for personal help there is paid support...
